core/src/state/world_state.rs

```diff
@@ -65,44 +66,94 @@ impl StateHandle {
             .write()
             .expect("Could not acquire world_state lock.")
     }
+}
+
+/*
+We want a thing that waits until notified, and importantly waits
+for that notify on all of its replicas.
+The default notifier from tokio does not behave the way we need
+since, if multiple clones are made, notify_one only causes the first
+to return immediately when it invokes the notification and awaits.
+notify_waiters does not cause even the first one to do that.
+Therefore, we implement our own notifier that can be closed and
+will return immediately on all its dependent notifies.
+*/
+#[derive(Debug)]
+pub struct ClosableNotify {
+    notifier: Arc<Sender<()>>,
+    rx: Receiver<()>,
+}
 
-    /** Block asynchronously until the given sequence number is reached. */
-    pub async fn wait_for_seq(&self, sequence: u64) {
-        if self.state().logical_time >= sequence {
-            return;
+impl ClosableNotify {
+    fn new() -> Self {
+        let (tx, rx) = tokio::sync::broadcast::channel(128);
+        Self {
+            notifier: Arc::new(tx),
+            rx,
         }
-        let receiver = { self.write_state().get_listener(sequence) };
-        receiver.notified().await;
+    }
+
+    pub async fn notified(&mut self) {
+        let _ = self.rx.recv().await;
+    }
+
+    fn notify_waiters(&self) {
+        let _ = self.notifier.send(());
     }
 }
 
+impl Clone for ClosableNotify {
+    fn clone(&self) -> Self {
+        Self {
+            notifier: self.notifier.clone(),
+            rx: self.notifier.subscribe(),
+        }
+    }
+}
 #[derive(Default, Debug)]
 pub struct WorldState {
     logical_time: u64,
     pub clusters: BTreeMap<ClusterName, ClusterState>,
-    listeners: HashMap<u64, Arc<Notify>>,
+    listeners: RwLock<BTreeMap<u64, ClosableNotify>>,
 }
 
 impl WorldState {
-    pub fn get_listener(&mut self, sequence: u64) -> Arc<Notify> {
-        match self.listeners.entry(sequence) {
-            Entry::Occupied(entry) => entry.get().clone(),
-            Entry::Vacant(entry) => {
-                let notify = Arc::new(Notify::new());
+    pub fn logical_time(&self) -> u64 {
+        self.logical_time
+    }
+
+    pub fn get_listener(&self, sequence: u64) -> ClosableNotify {
```

Comment:
Couldn't we simplify this and eliminate the need for `CloseableNotify` with something like:

```
pub async fn wait_for_seq(&self, sequence: u64) {
    if self.logical_time >= sequence {
        return;
    }
    match self.listeners.write().unwrap().entry(sequence) {
        Entry::Occupied(entry) => entry.get().subscribe().recv().await,
        Entry::Vacant(entry) => {
            let (send, recv) = broadcast::channel(1024);
            entry.insert(send);
            recv.recv().await
        }
    }
}
```

`self.listeners` would be a `BTreeMap<u64, Sender<()>>` and instead of calling `notify_waiters()` we would just call `.send(())` directly.

Reply:
And pass around sequence numbers to wait for instead of the notifies themselves? I guess that works.

The big reason for having closable notify is that it's clone, which makes it useful for the locking use case. Since I can shove it in a map and wait on it when needed.

Reply:
Hmm, on second thought, we really can't do the wait_for_seq thing the way you suggest since that'll require holding a lock to the statehandle, which won't work at all, unless I'm missing something. 

One thing I could do is to just use the normal notify if we want to keep complexity down in this code, and just have a doc comment telling people it's possible for it to deadlock if you don't await it in time.

Reply:
actually found a way to use a thinner wrapper around normal notify

Reply:
I had a go at it, let me know what you think of #424

Reply:
that looks good! For some reason I just wasn't thinking of future as just another trait. I've cherry-picked that commit onto this pr.